stablehlo/dialect/TypeInference.cpp

```diff
@@ -71,16 +71,65 @@ namespace {
 //===----------------------------------------------------------------------===//
 template <typename T>
 bool allQuantized(ArrayRef<Type> typeRange) {
-  return llvm::all_of(typeRange, [&](Type val) {
-    return val.cast<ShapedType>().getElementType().isa<T>();
-  });
+  return llvm::all_of(
+      typeRange, [&](Type val) { return getElementTypeOrSelf(val).isa<T>(); });
+}
+
+template <typename T>
+bool allQuantized(Type tp1, Type tp2) {
+  llvm::SmallVector<Type, 2> typeEntries{tp1, tp2};
+  return allQuantized<T>(typeEntries);
 }
 
 template <typename T>
 bool noneQuantized(ArrayRef<Type> typeRange) {
-  return llvm::all_of(typeRange, [&](Type val) {
-    return !val.cast<ShapedType>().getElementType().isa<T>();
-  });
+  return llvm::all_of(
+      typeRange, [&](Type val) { return !getElementTypeOrSelf(val).isa<T>(); });
+}
+
+template <typename QuantTy>
+bool isSameQuantScaleZeroPoint(Type ty1, Type ty2) {
+  auto qty1 = getElementTypeOrSelf(ty1).dyn_cast<QuantTy>();
+  auto qty2 = getElementTypeOrSelf(ty2).dyn_cast<QuantTy>();
+  if (!qty1 || !qty2) return false;
+
+  llvm::SmallVector<double> scales1;
+  llvm::SmallVector<double> scales2;
+  llvm::SmallVector<int64_t> zPoints1;
+  llvm::SmallVector<int64_t> zPoints2;
+
+  if (std::is_same<QuantTy, quant::UniformQuantizedType>::value) {
+    auto qty1 =
+        getElementTypeOrSelf(ty1).dyn_cast<quant::UniformQuantizedType>();
+    auto qty2 =
+        getElementTypeOrSelf(ty2).dyn_cast<quant::UniformQuantizedType>();
+    scales1.push_back(qty1.getScale());
+    scales2.push_back(qty2.getScale());
+    zPoints1.push_back(qty1.getZeroPoint());
+    zPoints2.push_back(qty2.getZeroPoint());
+  }
```

Comment:
`return ty1 == ty2` ?

Reply:
removed scale and zero_point matching for `per-tensor` quant. Instead, will be using `type1 == type2`. There is overlap of checks between HLO_CompatibleOperandsAndResultElementType Trait and using `type1 == type2` here. But it improves readability and reduces use of custom logic.